### ISSUE TITLE
The buffer length X does not match the expected buffer length for a picture of width W, height H and pixel depth D

### ISSUE THREAD

**Issue by seltix5** [yungd1plomat/AdvancedSharpAdbClient#3]
Hello,
I detected a memory leak on the current version of SharpAdbClient and I just created a issue there but the development does not appear to be very active.
Was this problem detected here since its a recent fork of that library?
Thanks!

**Comment by seltix5** [comment]
on the original library I had this problem, its present here too
https://github.com/quamotion/madb/issues/217
I will do the same here and test again.

**Comment by seltix5** [comment]
hello. I run some tests and the memory leak problem is caused by my "fix", I dont know why since it was the code used on the 2.2.41 version and there wasn't a memory leak there.
Any way, If I only replace the buffer.size with the calculation to get the suposed buffer size ()
https://github.com/yungd1plomat/AdvancedSharpAdbClient/blob/a0f04e62b332f1c31d4aadca15c26aa2fff1c83f/AdvancedSharpAdbClient/FramebufferHeader.cs#L165
and
https://github.com/yungd1plomat/AdvancedSharpAdbClient/blob/a0f04e62b332f1c31d4aadca15c26aa2fff1c83f/AdvancedSharpAdbClient/FramebufferHeader.cs#L227

and remove this check
https://github.com/yungd1plomat/AdvancedSharpAdbClient/blob/a0f04e62b332f1c31d4aadca15c26aa2fff1c83f/AdvancedSharpAdbClient/FramebufferHeader.cs#L191-L195

it does work as expected.

what do you think should be the best solution here?

**Comment by singaporeansimp** [comment]
Apologies, facing this issue myself now and came across this thread. Did y'all ever find a permanent solution?

**Comment by seltix5** [comment]
Hello, yes, the solution I posted lol
The buffer does not have a fixed size and thus can't be used to confirm the printscreen size, the only solution is to remove it or create a new variable to store that information but I don't think it's necessary. At least for me it's working without problems.

**Comment by singaporeansimp** [comment]
> Hello, yes, the solution I posted lol The buffer does not have a fixed size and thus can't be used to confirm the printscreen size, the only solution is to remove it or create a new variable to store that information but I don't think it's necessary. At least for me it's working without problems.

Forgive me, still a new C# developer. Made your edits to FramebufferHeader.cs and recompiled. Here's the code I'm using to fetch a screenshot:

```            
using (Task<Image> task = Client.GetFrameBufferAsync(Device, CancellationToken.None))
{
Image screenshot = task.GetAwaiter().GetResult();
return screenshot;
}```

**Comment by seltix5** [comment]
ohhh yes, there is multiple problems with the current version, it seams like the creator nor his users are using this library to get screenshots because there are serious problems and memory leak. I fixed those using code from the v2.2.41 where I knew there was no memory leaks. Here it is a list of my fixes :


**1 - Screenshot image is offset** (https://github.com/yungd1plomat/AdvancedSharpAdbClient/issues/5)
The size of the class will not give the correct size for the header data, replace this
https://github.com/yungd1plomat/AdvancedSharpAdbClient/blob/a0f04e62b332f1c31d4aadca15c26aa2fff1c83f/AdvancedSharpAdbClient/Framebuffer.cs#L50-L51
with this
```
this.headerData = new byte[52];
```


**2 - Bigger buffer than necessary**
The ArrayPool does not return precise size arrays, when used frequently it will allocate bigger memory than necessary.
This is the source problem of the checks referenced above because the Rent method returns a buffer size equal or bigger than requested (https://docs.microsoft.com/en-us/dotnet/api/system.buffers.arraypool-1.rent?view=net-5.0).
Replace this
https://github.com/yungd1plomat/AdvancedSharpAdbClient/blob/a0f04e62b332f1c31d4aadca15c26aa2fff1c83f/AdvancedSharpAdbClient/Framebuffer.cs#L117-L123
with this ( code from version 2.2.41 )
```
this.Data = new byte[this.Header.Size];
```
and remove this
https://github.com/yungd1plomat/AdvancedSharpAdbClient/blob/a0f04e62b332f1c31d4aadca15c26aa2fff1c83f/AdvancedSharpAdbClient/Framebuffer.cs#L152-L155
With this code update the checks referenced above (https://github.com/yungd1plomat/AdvancedSharpAdbClient/issues/3#issuecomment-943797492) will no longer be a problem since the buffer array will have the precise requested size and thus it will pass those checks.


**3 - Big memory leak**
I dont exactly know why because I dont know about sockets, but by trial and error, following the code sequence and removing code until I could pinpoint the source of the leak I found this
https://github.com/yungd1plomat/AdvancedSharpAdbClient/blob/a0f04e62b332f1c31d4aadca15c26aa2fff1c83f/AdvancedSharpAdbClient/SocketExtensions.cs#L54-L60
and replaced with the code from version 2.2.41
```
var cancellationTokenRegistration = cancellationToken.Register(() => socket.Dispose());

TaskCompletionSource<int> taskCompletionSource = new TaskCompletionSource<int>(socket);

socket.BeginReceive(buffer, offset, size, socketFlags, delegate (IAsyncResult iar)
{
    // this is the callback

    TaskCompletionSource<int> taskCompletionSource2 = (TaskCompletionSource<int>)iar.AsyncState;
    Socket socket2 = (Socket)taskCompletionSource2.Task.AsyncState;

    try
    {
        taskCompletionSource2.TrySetResult(socket2.EndReceive(iar));
    }
    catch (Exception ex)
    {
        if (ex is ObjectDisposedException && cancellationToken.IsCancellationRequested)
        {
            taskCompletionSource2.TrySetCanceled();
        }
        else
        {
            taskCompletionSource2.TrySetException(ex);
        }
    }
    finally
    {
        cancellationTokenRegistration.Dispose();
    }
}, taskCompletionSource);

return taskCompletionSource.Task;
```

**Comment by singaporeansimp** [comment]
Thanks @seltix5 -- looks like the memory leak issues are fixed now.
